networkx/algorithms/isomorphism/vf2pp.py

```diff
@@ -0,0 +1,276 @@
+
+Introduction
+------------
+The VF2++ algorithm, follows a similar logic to that of VF2, while also introducing new easy-to-check cutting rules and
+determining the optimal access order of nodes. It is also implemented in a non-recursive manner, which saves both time
+and space, when compared to its previous counterpart.
+
+The optimal node ordering is obtained after taking into consideration both the degree but also the label rarity of each
+node. This way we place the nodes that are more likely to match, first in the order, thus examining the most promising
+branches in the beginning. The rules also consider node labels, making it easier to prune unfruitful branches early in
+the process.
+
+Examples
+--------
+
+Suppose G1 and G2 are Isomorphic Graphs. Verification is as follows:
+
+Without node labels:
+
+>>> import networkx as nx
+>>> G1 = nx.path_graph(4)
+>>> G2 = nx.path_graph(4)
+>>> nx.vf2pp_is_isomorphic(G1, G2, node_labels=None)
+True
+>>> nx.vf2pp_mapping(G1, G2, node_labels=None)
+{1: 1, 2: 2, 0: 0, 3: 3}
+
+With node labels:
+
+>>> G1 = nx.path_graph(4)
+>>> G2 = nx.path_graph(4)
+>>> mapped = {1: 1, 2: 2, 3: 3, 0: 0}
+>>> nx.set_node_attributes(G1, dict(zip(G1, ["blue", "red", "green", "yellow"])), "label")
+>>> nx.set_node_attributes(G2, dict(zip([mapped[u] for u in G1], ["blue", "red", "green", "yellow"])), "label")
+>>> nx.vf2pp_is_isomorphic(G1, G2, node_labels="label")
+True
+>>> nx.vf2pp_mapping(G1, G2, node_labels="label")
+{1: 1, 2: 2, 0: 0, 3: 3}
+
+"""
+import collections
+
+import networkx as nx
+from networkx.algorithms.isomorphism.vf2pp_helpers.candidates import _find_candidates
+from networkx.algorithms.isomorphism.vf2pp_helpers.feasibility import _feasibility
+from networkx.algorithms.isomorphism.vf2pp_helpers.node_ordering import _matching_order
+from networkx.algorithms.isomorphism.vf2pp_helpers.state import (
+    _restore_Tinout,
+    _update_Tinout,
+)
+
+__all__ = ["vf2pp_mapping", "vf2pp_is_isomorphic", "vf2pp_all_mappings"]
+
+_GraphParameters = collections.namedtuple(
+    "_GraphParameters",
+    [
+        "G1",
+        "G2",
+        "G1_labels",
+        "G2_labels",
+        "nodes_of_G1Labels",
+        "nodes_of_G2Labels",
+        "G2_nodes_of_degree",
+    ],
+)
+_StateParameters = collections.namedtuple(
+    "_StateParameters",
+    ["mapping", "reverse_mapping", "T1", "T1_tilde", "T2", "T2_tilde"],
+)
+
+
+def vf2pp_mapping(G1, G2, node_labels=None, default_label=None):
+    """Return an isomorphic mapping between `G1` and `G2` if it exists.
+
+    Parameters
+    ----------
+    G1,G2: NetworkX Graph or MultiGraph instances.
+        The two graphs to check for isomorphism or monomorphism.
+
+    node_labels: Label name
+        The label name of all nodes
+
+    default_label: Label name
+        Let the user pick a default label value
+
+    Returns
+    -------
+    Node mapping, if the two graphs are isomorphic. None otherwise.
```

Comment:
Similar to the `Parameters` section, see the [docstring guide](https://numpydoc.readthedocs.io/en/latest/format.html#returns) for details on how the returns section should be formatted:

```suggestion
    dict or None
        Node mapping if the two graphs are isomorphic. None otherwise.
```